### DISCUSSION TITLE
help needed with this output

### DISCUSSION THREAD

**Discussion by J-HS-L** [microsoft/winget-cli#3091] [Q&A]
1Pwd, Spark, & Win Desktop Runtime, I already have those versions installed?!?
This seems to be happening a lot lately; when will winget be much more reliable, always so many bugbears
![Capture](https://user-images.githubusercontent.com/49305869/226306058-64c7f17f-ccf6-4729-bc2c-0b10cc0a2891.PNG)

And (I think) these are **all** other sections in the embedded ss above, that are confusing to this (old) *nix brain:

> [1]
> The following packages have an upgrade available, but require explicit targeting for upgrade:
> 
> [2]
> 1 package(s) have version numbers that cannot be determined. Use --include-unknown to see all results.
> 
> [3]
> Package download canceled.
> Cancelled
> (3a)
> 5 package(s) have version numbers that cannot be determined. Use --include-unknown to see all results.
> (3b)
> 1 package(s) are pinned and need to be explicitly upgraded.

Thanking you!

**Comment by Masamune3210** [comment]
Its up to the individual installers to update their versions in the registry, when they don't, this happens. Winget thinks that the versions arent installed when they really are. 
And, in order of appearance:
There is a package that requires you to explicitly call them as an upgrade target, usually because of something abnormal or unique about the package
There are packages that do not have known versions, this is caused when the installer for the package doesn't write the version of the app into the registry correctly. Since winget uses this information to know what is installed, without the version number it doesn't know what version is installed and by default hides them to prevent an upgrade loop scenario
Usually, when a package is canceled like this it is due to an error in my experience
Same as 2
You have a package that you have explicitly pinned, basically like held-back packages in something like apt. Since they are pinned, to upgrade them you have to explicitly call their package ID to upgrade them as a form of ensuring intention

**Comment by Masamune3210** [comment]
Ultimately it comes down to this. Winget is still relatively new in the package management scene, and its trying to wrangle a install base of programs that haven't really had to deal with adhering to the standards that securely or having something else seriously try to manage them, so there are a LOT of edge cases and bugs that have to be found, figured out, worked through the best solution that causes the least issues, and applied.
